### PR TITLE
Add support for replaying logs happen on async server operations

### PR DESCRIPTION
**Pull request by AbanoubGhadban**
### Summary

_Remove this paragraph and provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together._

### Pull Request checklist
_Remove this line after checking all the items here. If the item is not applicable to the PR, both check it out and wrap it by `~`._

- [ ] Add/update test to cover these changes
- [ ] Update documentation
- [ ] Update CHANGELOG file  
  _Add the CHANGELOG entry at the top of the file._

### Other Information

_Remove this paragraph and mention any other important and relevant information such as benchmarks._

- - -
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1649)

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved handling of `console.history` to ensure accurate console replay functionality across different execution contexts.
	- Enhanced management of `console.history` based on the rendering outcome, ensuring it resets appropriately.

- **New Features**
	- Added support for custom console history in console replay functionality, allowing for greater flexibility in console state management.
	- Introduced new types and functions to streamline the server rendering process for React components, enhancing error handling and output management.
	- Updated the `isPromise` function to handle a broader range of input types for improved flexibility.
	- Added new properties to the `RegisteredComponent` interface to improve component registration and rendering capabilities.
	- Initialized `console.history` as an empty array during server rendering for better console log management.

### CODE REVIEW

File: node_package/src/isServerRenderResult.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export function isServerRenderHash(testValue: CreateReactOutputResult):
     (testValue as ServerRenderResult).error);
 }
 
-export function isPromise(testValue: CreateReactOutputResult):
-  testValue is Promise<string> {
-  return !!((testValue as Promise<string>).then);
+export function isPromise<T>(testValue: CreateReactOutputResult | Promise<T> | string):
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply node_package/src/isServerRenderResult.ts]: Good suggestions. But when I tried this code
```
const testValue = '' as CreateReactOutputResult;
if (isPromise(testValue)) {
  testValue.then((y) => {
    console.log(y);
  });
}
```
I can see that typescript sees `testValue` as `Promise<string>` inside the `if` condition

File: node_package/src/serverRenderReactComponent.ts [hunk 2]
```diff
@@ -99,7 +98,6 @@ as a renderFunction and not a simple React Function Component.`);
     renderingError = e;
   }
 
-  const consoleReplayScript = buildConsoleReplay();
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: Updated

File: node_package/src/serverRenderReactComponent.ts [hunk 3]
```diff
@@ -112,8 +111,21 @@ as a renderFunction and not a simple React Function Component.`);
...
+        // If replayServerAsyncOperationLogs node renderer config is enabled, the console.history will contain all logs happened during sync and async operations.
+        // If the config is disabled, the console.history will be empty, because it will clear the history after the sync execution.
+        // In case of disabled config, we will use the console.history after sync execution, which contains all logs happened during sync execution.
+        const consoleHistoryAfterAsyncExecution = console.history;
+        let consoleReplayScript = '';
+        if ((consoleHistoryAfterAsyncExecution?.length ?? 0) > (consoleHistoryAfterSyncExecution?.length ?? 0)) {
+          consoleReplayScript = buildConsoleReplay(consoleHistoryAfterAsyncExecution);
+        } else {
+          consoleReplayScript = buildConsoleReplay(consoleHistoryAfterSyncExecution);
+        }
```
[the remark this answers is not in the document]
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: `console.history` getter can return different arrays with different calls.
Look here
https://github.com/shakacode/react_on_rails_pro/pull/440/files#diff-b1226881bc3c3c0b81e2590635966ca9671a4f98a9a2d2e2efa55294040c2db9R78
And here
https://github.com/shakacode/react_on_rails_pro/pull/440/files#diff-2d48090bb81968a3184f0218db6ec23dff44ec4c3b071c0be411fd177918257cR44

File: node_package/src/isServerRenderResult.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Judahmeek) [reply node_package/src/isServerRenderResult.ts]: Promise<string> is already included as one of the possible types of [CreateReactOutputResult](https://github.com/shakacode/react_on_rails/blob/master/node_package/src/types/index.ts#L41).

Should that definition just be expanded to Promise<T>?

Is string a new possible type?

File: node_package/src/serverRenderReactComponent.ts [hunk 4]
```diff
@@ -112,8 +111,21 @@ as a renderFunction and not a simple React Function Component.`);
...
+
         promiseResult = {
-          html: await renderResult,
+          html: awaitedRenderResult,
```
Comment (Judahmeek) [inline node_package/src/serverRenderReactComponent.ts]: Why are we basically just moving the `await` to line 114?

**Review by Judahmeek — APPROVED** [review]
Overall LGTM. Just questions abuot minor details.

**Comment by Judahmeek** [comment]
Actually, one additional question: Can we easily test these changes?

File: node_package/src/serverRenderReactComponent.ts [hunk 5]
```diff
@@ -99,7 +98,7 @@ as a renderFunction and not a simple React Function Component.`);
     renderingError = e;
   }
 
-  const consoleReplayScript = buildConsoleReplay();
+  const consoleHistoryAfterSyncExecution = console.history;
```
Comment (justin808) [inline node_package/src/serverRenderReactComponent.ts]: Too confusing to use this getter...timing and globals

Could this be updated only for the pro version node renderer to return a object for the console history?

https://github.com/shakacode/react_on_rails/pull/1649/files#diff-cbb0f0518343f42b78a0ec373eddd1187ffb9275651ddb9758a1160c52176eb6R28-R34

```js
    const reactRenderingResult = createReactOutput({
      componentObj,
      domNodeId,
      trace,
      props,
      railsContext,
    });
```

File: node_package/src/serverRenderReactComponent.ts (hunk 4 above)
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: The console replay script must be built after awaiting the rendering promise, so it includes all console logs happened during sync and async operations on the server.

File: node_package/src/serverRenderReactComponent.ts (hunk 5 above)
Comment (AbanoubGhadban) [reply node_package/src/serverRenderReactComponent.ts]: It's updated. Implementation is simpler now.

**Pull request merged**